libraries/AP_ESC_Telem/AP_ESC_Telem.cpp

```diff
@@ -105,14 +105,12 @@ uint8_t AP_ESC_Telem::get_motor_frequencies_hz(uint8_t nfreqs, float* freqs) con
 uint32_t AP_ESC_Telem::get_active_esc_mask() const {
     uint32_t ret = 0;
     const uint32_t now = AP_HAL::millis();
-    uint32_t now_us = AP_HAL::micros();
     for (uint8_t i = 0; i < ESC_TELEM_MAX_ESCS; i++) {
         if (_telem_data[i].last_update_ms == 0 && !was_rpm_data_ever_reported(_rpm_data[i])) {
             // have never seen telem from this ESC
             continue;
         }
-        if (_telem_data[i].stale(now)
-            && !rpm_data_within_timeout(_rpm_data[i], now_us, ESC_RPM_DATA_TIMEOUT_US)) {
+        if (_telem_data[i].stale(now) && !_rpm_data[i].data_valid) {
```

Comment:
```suggestion
        if (_telem_data[i].stale(now) || !_rpm_data[i].data_valid) {
```
If data stale or not valid then don't used it?

Reply:
That would be a change to the current behavior. I don't understand why, but this logic was very clearly intentional. The function's description even mentions "and/or" instead of "and"

Reply:
It seems to be because there are two timeouts. One for "everything else" and one for "RPM". So the assumption is that if we got "everything else" recently then a timed out RPM is better than nothing. I'm not sure that really makes sense, it would mean we use RPM from more than one second ago. I wonder if this is a bodge to make up for the issue your fixing.

Since we also have a timeout for "everything else" I think we also need a fix the same as the one you have done for RPM for it?

Reply:
> we use RPM from more than one second ago

Not really. This logic is only in `get_active_esc_mask` and `get_max_rpm_esc`, the latter is only used for OSD. `get_rpm` checks the timeout of RPM only, not the other telem, so we don't have an issue. Arguably, the "other telem" timeout check is unnecessary in `get_max_rpm_esc` and should be dropped entirely, not just changed to `||`.

> Since we also have a timeout for "everything else" I think we also need a fix the same as the one you have done for RPM for it?

Yup: https://github.com/ArduPilot/ardupilot/pull/29028

Reply:
OK, its a bit odd but I'm convinced. Thanks!